src/qutip_cupy/dense_functions.py

```diff
@@ -29,3 +29,30 @@ def trace_cupydense(cp_arr):
     # @TODO: whnen qutip allows it we should remove this call to item()
     # as it takes a time penalty commmunicating data from GPU to CPU.
     return cp.trace(cp_arr._cp).item()
+
+
+def isherm_cupydense(cp_arr, tol):
+    if cp_arr.shape[0] != cp_arr.shape[1]:
+        return False
+    size = cp_arr.shape[0]
+    hermcheck_kernel = cp.RawKernel(
+        r"""
+    #include <cupy/complex.cuh>
+    extern "C" __global__
+    void hermcheck(const complex<double>* x1,const int size,
+                    const double tol, unsigned int* y) {
+      int tidx = blockDim.x * blockIdx.x + threadIdx.x;
+      int tidy = blockDim.y * blockIdx.y + threadIdx.y;
+
+    atomicAnd(y,int(norm(x1[tidx*size+tidy] - conj(x1[tidy*size+tidx])) < tol));
```

Comment:
It'd be interesting to compare the atomic approach vs a reduction-based approach. In the old days of CUDA atomic operations were not very performant, leading to development of parallel algorithms as an alternative. However, on newer architectures and/or drivers maybe things have improved quite a bit.

In this particular case, I think you can create an intermediate buffer to hold the results, and then call a reduction like `cp.all()` to complete the check. This has an additional benefit that you can achieve it using standard CuPy ndarray manipulations without any custom CUDA kernel. But ofc you need to do a careful benchmark.

Reply:
I read again and my guts would go with reduction, as `atomicAdd` would perform worse as the matrix size gets larger (O(N^2) calls to atomicAdd that serialize the operations).

Reply:
![atomic vs cupy reduction](https://user-images.githubusercontent.com/23513768/128258225-6c6a3392-a168-4eb0-bd76-1f7f637d246a.png)

Looks like we have a winner

Reply:
Nice! Thanks Felipe.